File: src/resilience/circuit_breaker.py
#!/usr/bin/env python3
"""
Circuit breaker pattern for Hedera and external API resilience.

Prevents cascade failures when Hedera Mirror Node or CoinGecko are down.
States: CLOSED (normal) -> OPEN (failing) -> HALF_OPEN (testing recovery)

Usage:
    breaker = CircuitBreaker(failure_threshold=5, recovery_timeout=30)

    @breaker
    def fetch_hedera_data():
        return requests.get("https://mainnet-public.mirrornode.hedera.com/...")

    result = fetch_hedera_data()  # Automatically protected
"""

import time
import functools
from enum import Enum
from typing import Any, Callable, Optional, Dict


__all__ = [
    "CircuitState", "CircuitBreaker", "CircuitBreakerOpenError",
    "APICircuitBreakers", "retry_with_backoff",
]


class CircuitState(Enum):
    CLOSED = "closed"      # Normal operation
    OPEN = "open"          # Failing, reject fast
    HALF_OPEN = "half_open"  # Testing if service recovered


class CircuitBreaker:
    """
    Circuit breaker for external API calls.

    Args:
        failure_threshold: Number of failures before opening circuit
        recovery_timeout: Seconds to wait before trying again
        half_open_max_calls: Number of test calls in half-open state
        success_threshold: Successes needed to close circuit
    """

    def __init__(
        self,
        name: str = "default",
        failure_threshold: int = 5,
        recovery_timeout: float = 30.0,
        half_open_max_calls: int = 3,
        success_threshold: int = 2,
        fallback_value: Any = None,
    ):
        self.name = name
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.half_open_max_calls = half_open_max_calls
        self.success_threshold = success_threshold
        self.fallback_value = fallback_value

        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time = 0
        self.half_open_calls = 0

        # Metrics
        self.total_calls = 0
        self.total_failures = 0
        self.total_successes = 0
        self.total_fallbacks = 0
        self.state_changes = []

    def _transition_to(self, new_state: CircuitState):
        """Record state transition."""
        old_state = self.state
        self.state = new_state
        self.state_changes.append({
            "from": old_state.value,
            "to": new_state.value,
            "timestamp": time.time(),
        })

        # Reset counters on transition
        if new_state == CircuitState.CLOSED:
            self.failure_count = 0
            self.success_count = 0
            self.half_open_calls = 0
        elif new_state == CircuitState.OPEN:
            self.success_count = 0
            self.half_open_calls = 0
        elif new_state == CircuitState.HALF_OPEN:
            self.failure_count = 0
            self.success_count = 0
            self.half_open_calls = 0

    def _should_attempt(self) -> bool:
        """Check if we should attempt the call based on current state."""
        if self.state == CircuitState.CLOSED:
            return True

        if self.state == CircuitState.OPEN:
            # Check if recovery timeout has passed
            if time.time() - self.last_failure_time >= self.recovery_timeout:
                self._transition_to(CircuitState.HALF_OPEN)
                return True
            return False

        if self.state == CircuitState.HALF_OPEN:
            if self.half_open_calls < self.half_open_max_calls:
                return True
            return False

        return True

    def _record_success(self):
        """Record a successful call."""
        self.total_successes += 1

        if self.state == CircuitState.HALF_OPEN:
            self.success_count += 1
            self.half_open_calls += 1

            if self.success_count >= self.success_threshold:
                self._transition_to(CircuitState.CLOSED)
        else:
            self.failure_count = 0  # Reset failure count on success

    def _record_failure(self):
        """Record a failed call."""
        self.total_failures += 1
        self.failure_count += 1
        self.last_failure_time = time.time()

        if self.state == CircuitState.HALF_OPEN:
            self.half_open_calls += 1
            self._transition_to(CircuitState.OPEN)
        elif self.failure_count >= self.failure_threshold:
            self._transition_to(CircuitState.OPEN)

    def __call__(self, func: Callable) -> Callable:
        """Decorator to wrap a function with circuit breaker."""
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            return self.call(func, *args, **kwargs)

        # Attach breaker for inspection
        wrapper._circuit_breaker = self
        return wrapper

    def call(self, func: Callable, *args, **kwargs) -> Any:
        """Execute function with circuit breaker protection."""
        self.total_calls += 1

        if not self._should_attempt():
            self.total_fallbacks += 1
            if self.fallback_value is not None:
                return self.fallback_value
            raise CircuitBreakerOpenError(
                f"Circuit '{self.name}' is OPEN. Last failure: {self.last_failure_time}"
            )

        try:
            result = func(*args, **kwargs)
            self._record_success()
            return result
        except Exception as e:
            self._record_failure()
            raise

    def get_state(self) -> Dict[str, Any]:
        """Get current circuit breaker state."""
        return {
            "name": self.name,
            "state": self.state.value,
            "failure_count": self.failure_count,
            "success_count": self.success_count,
            "last_failure_time": self.last_failure_time,
            "total_calls": self.total_calls,
            "total_failures": self.total_failures,
            "total_successes": self.total_successes,
            "total_fallbacks": self.total_fallbacks,
            "failure_rate": round(self.total_failures / max(self.total_calls, 1) * 100, 2),
        }


class CircuitBreakerOpenError(Exception):
    """Raised when circuit breaker is open."""
    pass


# Pre-configured breakers for common APIs
class APICircuitBreakers:
    """Circuit breakers for all external APIs."""

    def __init__(self):
        # Hedera Mirror Node: tolerant (can be slow, but don't hammer it)
        self.hedera_mirror = CircuitBreaker(
            name="hedera_mirror",
            failure_threshold=3,
            recovery_timeout=60.0,
            fallback_value={"error": "Hedera Mirror Node unavailable", "cached": True},
        )

        # CoinGecko: strict (rate limits, don't retry too aggressively)
        self.coingecko = CircuitBreaker(
            name="coingecko",
            failure_threshold=2,
            recovery_timeout=120.0,
            fallback_value={"error": "CoinGecko rate limited", "cached": True},
        )

        # Hedera Consensus Service: moderate
        self.hcs = CircuitBreaker(
            name="hcs",
            failure_threshold=5,
            recovery_timeout=30.0,
            fallback_value={"error": "HCS temporarily unavailable", "cached": True},
        )

        # Hedera Token Service: moderate
        self.hts = CircuitBreaker(
            name="hts",
            failure_threshold=5,
            recovery_timeout=30.0,
            fallback_value={"error": "HTS temporarily unavailable", "cached": True},
        )

    def get_all_states(self) -> Dict[str, Any]:
        """Get all circuit breaker states."""
        return {
            "hedera_mirror": self.hedera_mirror.get_state(),
            "coingecko": self.coingecko.get_state(),
            "hcs": self.hcs.get_state(),
            "hts": self.hts.get_state(),
        }

    def reset_all(self):
        """Reset all circuit breakers to CLOSED."""
        for breaker in [self.hedera_mirror, self.coingecko, self.hcs, self.hts]:
            breaker._transition_to(CircuitState.CLOSED)


# Retry with exponential backoff
def retry_with_backoff(
    max_retries: int = 3,
    base_delay: float = 1.0,
    max_delay: float = 60.0,
    exceptions: tuple = (Exception,),
    on_retry: Callable = None,
):
    """Retry decorator with exponential backoff and jitter."""
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            delay = base_delay

            for attempt in range(max_retries + 1):
                try:
                    return func(*args, **kwargs)
                except exceptions as e:
                    if attempt == max_retries:
                        raise

                    # Exponential backoff with jitter
                    import random
                    jitter = random.uniform(0, delay * 0.1)
                    sleep_time = min(delay + jitter, max_delay)

                    if on_retry:
                        on_retry(attempt + 1, max_retries, e, sleep_time)

                    time.sleep(sleep_time)
                    delay *= 2

            # Should never reach here
            raise RuntimeError("Retry loop exited unexpectedly")

        return wrapper
    return decorator


if __name__ == "__main__":
    # Test circuit breaker
    breaker = CircuitBreaker(
        name="test",
        failure_threshold=3,
        recovery_timeout=5.0,
    )

    @breaker
    def flaky_function(should_fail: bool = False):
        if should_fail:
            raise RuntimeError("Simulated failure")
        return "success"

    # Test normal operation
    print("Test 1: Normal call")
    print(f"  Result: {flaky_function(False)}")
    print(f"  State: {breaker.get_state()}")

    # Test failures
    print("\nTest 2: Trigger failures")
    for i in range(5):
        try:
            flaky_function(True)
        except Exception as e:
            print(f"  Call {i+1}: {e}")

    print(f"  State: {breaker.state.value}")
    print(f"  Stats: {breaker.get_state()}")

    # Test circuit open
    print("\nTest 3: Circuit should be open")
    try:
        flaky_function(False)
    except CircuitBreakerOpenError as e:
        print(f"  Expected: {e}")

    # Test recovery
    print("\nTest 4: Wait for recovery")
    time.sleep(6)
    print(f"  State after timeout: {breaker.state.value}")

    # Test half-open recovery
    for i in range(3):
        try:
            result = flaky_function(False)
            print(f"  Recovery call {i+1}: {result}")
        except Exception as e:
            print(f"  Recovery call {i+1}: {e}")

    print(f"  Final state: {breaker.state.value}")
    print(f"  Final stats: {breaker.get_state()}")

File: src/vera/proofKernel/circuitBreaker.ts
/**
 * Circuit Breaker Pattern for Meridian HTTP Clients
 * 
 * Prevents cascade failures by opening circuit after threshold failures
 * Automatically attempts recovery after cooldown period
 */

export type CircuitState = 'closed' | 'open' | 'half-open';

export interface CircuitBreakerConfig {
  failureThreshold: number;      // Failures before opening circuit
  resetTimeoutMs: number;        // Time before attempting recovery
  halfOpenMaxCalls: number;      // Test calls in half-open state
  successThreshold: number;      // Successes required to close circuit
}

export interface CircuitBreakerStats {
  state: CircuitState;
  failures: number;
  successes: number;
  lastFailureTime: number | null;
  lastSuccessTime: number | null;
  consecutiveSuccesses: number;
  consecutiveFailures: number;
  totalCalls: number;
  rejectedCalls: number;
}

const DEFAULT_CONFIG: CircuitBreakerConfig = {
  failureThreshold: 3,
  resetTimeoutMs: 30000,  // 30 seconds
  halfOpenMaxCalls: 3,
  successThreshold: 2,
};

export class CircuitBreaker {
  private state: CircuitState = 'closed';
  private stats: CircuitBreakerStats = {
    state: 'closed',
    failures: 0,
    successes: 0,
    lastFailureTime: null,
    lastSuccessTime: null,
    consecutiveSuccesses: 0,
    consecutiveFailures: 0,
    totalCalls: 0,
    rejectedCalls: 0,
  };
  private halfOpenCalls = 0;
  private nextAttempt = 0;

  constructor(
    private readonly name: string,
    private readonly config: CircuitBreakerConfig = DEFAULT_CONFIG,
  ) {}

  getState(): CircuitState {
    this.updateState();
    return this.state;
  }

  getStats(): CircuitBreakerStats {
    return { ...this.stats, state: this.state };
  }

  canExecute(): boolean {
    this.updateState();

    if (this.state === 'open') {
      this.stats.rejectedCalls++;
      return false;
    }

    if (this.state === 'half-open' && this.halfOpenCalls >= this.config.halfOpenMaxCalls) {
      this.stats.rejectedCalls++;
      return false;
    }

    return true;
  }

  recordSuccess(): void {
    const now = Date.now();
    this.stats.successes++;
    this.stats.consecutiveSuccesses++;
    this.stats.consecutiveFailures = 0;
    this.stats.lastSuccessTime = now;
    this.stats.totalCalls++;

    if (this.state === 'half-open') {
      this.halfOpenCalls++;
      
      if (this.stats.consecutiveSuccesses >= this.config.successThreshold) {
        this.closeCircuit();
      }
    }
  }

  recordFailure(error?: Error): void {
    const now = Date.now();
    this.stats.failures++;
    this.stats.consecutiveFailures++;
    this.stats.consecutiveSuccesses = 0;
    this.stats.lastFailureTime = now;
    this.stats.totalCalls++;

    if (this.state === 'half-open') {
      this.openCircuit();
      return;
    }

    if (this.state === 'closed' && this.stats.consecutiveFailures >= this.config.failureThreshold) {
      this.openCircuit();
    }
  }

  private updateState(): void {
    const now = Date.now();

    if (this.state === 'open' && now >= this.nextAttempt) {
      this.state = 'half-open';
      this.halfOpenCalls = 0;
      console.log(`[CircuitBreaker:${this.name}] Entering half-open state - testing recovery`);
    }
  }

  private openCircuit(): void {
    this.state = 'open';
    this.nextAttempt = Date.now() + this.config.resetTimeoutMs;
    this.halfOpenCalls = 0;
    console.log(`[CircuitBreaker:${this.name}] Circuit OPENED - cooling down for ${this.config.resetTimeoutMs}ms`);
  }

  private closeCircuit(): void {
    this.state = 'closed';
    this.stats.consecutiveFailures = 0;
    this.halfOpenCalls = 0;
    console.log(`[CircuitBreaker:${this.name}] Circuit CLOSED - healthy`);
  }

  async execute<T>(fn: () => Promise<T>, fallback?: () => T): Promise<T> {
    if (!this.canExecute()) {
      if (fallback) {
        console.log(`[CircuitBreaker:${this.name}] Circuit open, using fallback`);
        return fallback();
      }
      throw new Error(`Circuit breaker ${this.name} is open`);
    }

    try {
      const result = await fn();
      this.recordSuccess();
      return result;
    } catch (error) {
      this.recordFailure(error instanceof Error ? error : new Error(String(error)));
      throw error;
    }
  }
}

// Circuit breaker registry for managing multiple breakers
export class CircuitBreakerRegistry {
  private breakers = new Map<string, CircuitBreaker>();

  getOrCreate(name: string, config?: CircuitBreakerConfig): CircuitBreaker {
    if (!this.breakers.has(name)) {
      this.breakers.set(name, new CircuitBreaker(name, config));
    }
    return this.breakers.get(name)!;
  }

  getStats(): Record<string, CircuitBreakerStats> {
    const stats: Record<string, CircuitBreakerStats> = {};
    for (const [name, breaker] of this.breakers) {
      stats[name] = breaker.getStats();
    }
    return stats;
  }

  healthCheck(): { healthy: boolean; details: Record<string, CircuitState> } {
    const details: Record<string, CircuitState> = {};
    let healthy = true;

    for (const [name, breaker] of this.breakers) {
      const state = breaker.getState();
      details[name] = state;
      if (state === 'open') healthy = false;
    }

    return { healthy, details };
  }
}

export const globalCircuitBreakers = new CircuitBreakerRegistry();
